libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp

```diff
@@ -0,0 +1,303 @@
+        case SDP3X_SCALE_PRESSURE_SDP33:
+            c = '3';
+            break;
+        }
+        hal.console->printf("SDP3%c: Found on bus %u address 0x%02x scale=%u\n",
+                            c, get_bus(), addresses[i], _scale);
+    }
+
+    if (!found) {
+        return false;
+    }
+
+    /*
+      this sensor uses zero offset and skips cal
+     */
+    set_use_zero_offset();
+    set_skip_cal();
+    set_offset(0);
+    
+    // drop to 2 retries for runtime
+    _dev->set_retries(2);
+
+    _dev->register_periodic_callback(20000,
+                                     FUNCTOR_BIND_MEMBER(&AP_Airspeed_SDP3X::_timer, void));
+    return true;
+}
+
+// read the values from the sensor. Called at 50Hz
+void AP_Airspeed_SDP3X::_timer()
+{
+    // read 6 bytes from the sensor
+    uint8_t val[6];
+    int ret = _dev->transfer(nullptr, 0, &val[0], sizeof(val));
+    if (!ret) {
+        return;
+    }
+    // Check the CRC
+    if (!_crc(&val[0], 2, val[2]) || !_crc(&val[3], 2, val[5])) {
+        return;
+    }
+
+    int16_t P = (((int16_t)val[0]) << 8) | val[1];
+    int16_t temp = (((int16_t)val[3]) << 8) | val[4];
+
+    float P_float = (float)P;
+    float temp_float = (float)temp;
+
+    float scale_float = (float)_scale;
+
+    float diff_press_pa = P_float / scale_float;
+    float temperature = temp_float / (float)SDP3X_SCALE_TEMPERATURE;
+
+    if (sem->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        _press_sum += diff_press_pa;
+        _temp_sum += temperature;
+        _press_count++;
+        _temp_count++;
+        _last_sample_time_ms = AP_HAL::millis();
+        sem->give();
+    }
+}
+
+/*
+  correct pressure for barometric height
+  With thanks to:
+   https://github.com/PX4/Firmware/blob/master/Tools/models/sdp3x_pitot_model.py
+ */
+float AP_Airspeed_SDP3X::_correct_pressure(float press)
+{
+    const float tube_len = 0.2;
+    float temperature;
+    AP_Baro *baro = AP_Baro::get_instance();
+
+    if (baro == nullptr) {
+        return press;
+    }
+
+    // fix for tube order
+    AP_Airspeed::pitot_tube_order tube_order = get_tube_order();
+    switch (tube_order) {
+    case AP_Airspeed::PITOT_TUBE_ORDER_NEGATIVE:
+        press = -press;
+    //FALLTHROUGH;
+    case AP_Airspeed::PITOT_TUBE_ORDER_POSITIVE:
+        break;
+    case AP_Airspeed::PITOT_TUBE_ORDER_AUTO:
+    default:
+        press = fabsf(press);
+        break;
+    }
+    if (press <= 0) {
+        return 0;
+    }
+
+    get_temperature(temperature);
+    float rho_air = baro->get_pressure() / (CONSTANTS_AIR_GAS_CONST * (temperature - CONSTANTS_ABSOLUTE_NULL_CELSIUS));
+
+    // flow through sensor
+    float flow_SDP3X = (300.805f - 300.878f / (0.00344205f * (float)powf(press, 0.68698f) + 1)) * 1.29f / rho_air;
+    if (flow_SDP3X < 0.0f) {
+        flow_SDP3X = 0.0f;
+    }
+
+    // diffential pressure through pitot tube
+    float dp_pitot = 28557670.0f - 28557670.0f / (1 + (float)powf((flow_SDP3X / 5027611.0f), 1.227924f));
```

Comment:
Mystery constants are mysterious!

Reply:
Looks like they came from a calibration using this method: https://www.sensirion.com/fileadmin/user_upload/customers/sensirion/Dokumente/8_Differential_Pressure/Sensirion_Differential_Pressure_Sensors_SDP3x_Pressure_Drop_In_Hose.pdf

The weird thing about these sesnors is that they measure differential pressure by measuring mass flow through a calibrated orifice. Unlike a traditional membrane sensor this is not a static measurement, so the head loss in the tubing must be accounted for.